storage/tsdb/tsdb.go

```diff
@@ -138,6 +146,47 @@ func Open(path string, l log.Logger, r prometheus.Registerer, opts *Options) (*t
 	return db, nil
 }
 
+func (a adapter) StartTime() (int64, error) {
+	startTime := int64(model.Latest)
+
+	var indexr tsdb.IndexReader
+	if len(a.db.Blocks()) > 0 {
+		indexr = a.db.Blocks()[0].Index()
+	} else {
+		indexr = a.db.Head().Index()
+	}
+
+	joblabel := "job"
+	tpls, err := indexr.LabelValues(joblabel)
+	if err != nil {
+		return startTime, err
+	}
+
+	for i := 0; i < tpls.Len(); i++ {
```

Comment:
I think it would much simpler to just return `db.Blocks()[0].MinTime` (or `head.MinTime`). You will gain nothing by going through each series as while the `series.MinTime` may differ from the `block.MinTime`, if the primary has no data, then the remote too won't have it.

Reply:
There is some situations where the primary has no data and the remote do have data. For example, if you use prometheus with an ephemeral disk (like running prometheus on marathon/mesos/kube/borg), and restart your prometheus instance you loose local data, but remote do still have data.
That's why it's important to have a more precise startTime than `block.MinTime`.

Reply:
Umm, yes, but when you restart your prometheus, your `block.MinTime` would essentially change to reflect the data you have. Like if you start with an empty disk, it will be come `time.Now()` (or the first that of the sample appended). For example if you have three blocks with min times: 10, 20, 30, and you lose the first block, the `db.Blocks().MinTime` would now reflect 20 not 10.

As I see it, this is extra complexity that has no benefit.

Reply:
As far as I understand, there is nothing that links `block.MinTime` to the very first timestamp in the `DB` or `time.now()`.
The goal here is to avoid useless remote read calls. This means checking if some time range for which the samples are not locally present. And if there is some, limit the remote read query to a `maxt` that reduce overlap between local samples and remote samples.

My current proposal is to use maxt = localStartTime (no overlap at all). Then we need to have enough accuracy for it's value to avoid gaps. Example of gaps :
- The default HEAD block width is 2h (`storage.tsdb.min-block-duration`). This mean that after a restart with empty disk the difference between `block.MinTime` and the very first timestamp is between 0s and 2h. So most of the time we get gaps in timeseries that are much larger than a few minutes.
- But this is even worst if local storage keep alive until it fill samples for the entire retention duration. Indeed, the default `MaxBlockDuration` is `10% of the retention period`. This mean that after some time, the difference `block.MinTime` and the very first timestamp is 10% of the retention. This would lead to even larger gap in timeseries.
Therefore, from these point of view these extra complexity do have benefit.

But we could choose an other point of view with a second solution : using `block.MinTime` as localStartTime (as you suggest), and `block.MaxTime` as `maxt`.
- no extra complexity anymore, and no gap
- but we would still have some inefficient queries since we would query over time range that are partially already stored in local tsdb (limited to queries that retrieve points from the first block, so not all queries)

At first, I though that the extra complexity would not affect prometheus performances (but was not able to quantify it) and would better represents stored timestamps and remove all useless remote read calls. But it's not the only solution.
So if you think the second solution is more viable, I'm ok. This is still an improvement since it avoid most of useless remote read calls.
What do you think, @Gouthamve

Reply:
> This mean that after a restart with empty disk the difference between block.MinTime and the very first timestamp is between 0s and 2h. So most of the time we get gaps in timeseries that are much larger than a few minutes.

Makes sense, I read the code again to find that it actually could be a `2h` diff! Until now I was thinking `MinTime` would be that of the earliest sample in the db. Ref: https://github.com/prometheus/tsdb/blob/963a270885c7d8b2cc3dafbe0e44b36ff031ca56/head.go#L296

Sorry for the noise!

Reply:
Ok no pb, it's fine for me